service/s3/presign_post.go

```diff
@@ -0,0 +1,408 @@
+package s3
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/aws-sdk-go-v2/aws/retry"
+	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+	internalcontext "github.com/aws/aws-sdk-go-v2/internal/context"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	acceptencodingcust "github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding"
+	presignedurlcust "github.com/aws/aws-sdk-go-v2/service/internal/presigned-url"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+)
+
+const (
+	algorithmHeader  = "X-Amz-Algorithm"
+	credentialHeader = "X-Amz-Credential"
+	dateHeader       = "X-Amz-Date"
+	tokenHeader      = "X-Amz-Security-Token"
+	signatureHeader  = "X-Amz-Signature"
+
+	algorithm        = "AWS4-HMAC-SHA256"
+	aws4Request      = "aws4_request"
+	bucketHeader     = "bucket"
+	defaultExpiresIn = 15 * time.Minute
+)
+
+// PresignedPostRequest represents a presigned request to be sent using HTTP verb POST and FormData
+type PresignedPostRequest struct {
+	// Represents the Base URL to make a request to
+	URL string
+	// Values is a key-value map of values to be sent as FormData
+	// these values are not encoded
+	Values map[string]interface{}
```

Comment:
Can this just be map[string]string? Since the final result is just strings for the POST form, right?

Reply:
You are right. Was getting tripped with the other JSON values having to be map[string]interface{}, but this can actually be just as string to string. Good catch!